lib/streamlit/caching.py

```diff
@@ -499,6 +499,16 @@ def cache(
             hash_funcs=hash_funcs,
         )
 
+    name = func.__name__
+
+    initial_hasher = hashlib.new("md5")
```

Comment:
The effect is that we're only caching the explicit argument values and the (unexecuted) function body and its references.

This way, the only time we actually execute and evaluate the function is when there's a cache MISS.

The intended effect is that we disinclude the _values_ of closure variables and referenced functions from the evaluation of whether something has changed in the cache, without having to change anything about how the internals of hashing work at all.

It's quite a big behavior change, because we're shifting from something slightly more magical in which the cache recognizes that implicit inputs have changed, to something a lot more user-legible in which the cache treats its inputs more deterministically.

That is, from a cached function's perspective the way the world works is that a global doesn't change, and a referenced function's output for a given argument doesn't change unless its _code_ has changed.

Reply:
I get the high level of what the PR is doing but specifically what's the effect of moving lines 502 - 510 into their place instead of keeping them where they were below?

Reply:
This was spurious code that popped in again during remote merge...